### ISSUE TITLE
-Wconversion triggering errors when using gcc-7.1

### ISSUE THREAD

**Issue by cpockrandt** [ericniebler/range-v3#680]
Hi,

I moved from using the latest gcc-6 to gcc-7.1 and noticed that the additional flag `-Wconversion` causes a couple of errors in your library which worked perfectly fine with previous gcc versions.

Do you unofficially support this flag or was it kind of a "happy accident" that it used to work before? 😃 All tests succeeded and with the newest gcc-7 only a couple are failing with the same error message.

Here's a minimal example:

    std::vector<char> xs {'A', 'B'};
    std::vector<int> ys {1, 2};
    for (auto [ x, y ] : ranges::view::zip(xs, ys))
        std::cout << x << ", " << y << std::endl;

Output:

    ../range-v3/include/range/v3/utility/common_type.hpp:76:45: warning: choosing ‘ranges::v3::common_pair<F, S>::operator std::pair<_U1, _U2>() const & [with F2 = char; S2 = int; int _concept_requires_276 = 42; typename std::enable_if<((_concept_requires_276 == 43) || (ranges::v3::concepts::models<ranges::v3::concepts::Constructible, F2, const F&>() && ranges::v3::concepts::models<ranges::v3::concepts::Constructible, S2, const S&>())), int>::type <anonymous> = 0; F = char&; S = int&]’ over ‘constexpr std::pair<_T1, _T2>::pair(const std::pair<_U1, _U2>&) [with _U1 = char&; _U2 = int&; typename std::enable_if<(std::_PCC<((! std::is_same<_T1, _U1>::value) || (! std::is_same<_T2, _U2>::value)), _T1, _T2>::_ConstructiblePair<_U1, _U2>() && std::_PCC<((! std::is_same<_T1, _U1>::value) || (! std::is_same<_T2, _U2>::value)), _T1, _T2>::_ImplicitlyConvertiblePair<_U1, _U2>()), bool>::type <anonymous> = 1; _T1 = char; _T2 = int]’ [-Wconversion]
             using _cond_res = decltype(true ? std::declval<T>() : std::declval<U>());
                                        ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    ../range-v3/include/range/v3/utility/common_type.hpp:76:45: warning:   for conversion from ‘’ to ‘const std::pair<char, int>’ [-Wconversion]
    ../range-v3/include/range/v3/utility/common_type.hpp:76:45: note:   because conversion sequence for the argument is better

**Comment by CaseyCarter** [comment]
Happy accident As a rule, any warning that `CMakeLists.txt` doesn't enable for the tests is a warning we don't support.

The range-v3 diagnostic story is a little poor right now. I plan to improve the situation by changing our CI configuration to turn on every possible warning, and individually suppress the warnings we don't care about in the headers, but it's blocked waiting on the CMake overhaul in progress in #600.

This particular occurrence is probably due to the common type of `common_pair` and its base `std::pair` being `std::pair`, which I - and apparently GCC 7 - think is a bit weird.

**Comment by h-2** [comment]
It would be great if tests could include `-Wall -Wextra -Wconversion` for GCC and at least `-Wall -Wshorten-64-to-32` on Clang as these warnings prevent people from writing unsafe code! And depending on a library that doesn't enforce them makes it harder for downstream projects...
Thanks!

**Comment by ericniebler** [comment]
> This particular occurrence is probably due to the common type of `common_pair` and its base `std::pair` being `std::pair`, which I - and apparently GCC 7 - think is a bit weird.

Why is that weird? Makes sense to me considering substitutability.

**Comment by CaseyCarter** [comment]
> Why is that weird?

Derived types are nominally more-specific than their base types since they add behavior. If the common type of `Derived` and `Base` is `Base`, it suggests that the conversion from `Derived` to `Base` loses no information. If that is the case, then they must be equivalent - why define `Derived` at all? If they are *not* equivalent, then we are cheating `Common`.

**Comment by ericniebler** [comment]
Any code that works with a `Base` should work with a `Derived`. Consider:

```
Base x = Base{};    // OK
Base y = Derived{}; // OK
```

Now consider this:

```
Derived x = Base{};    // OOPS!
Derived y = Derived{}; // OK
```

`Derived` cannot be the common type between `Derived` and `Base` because that's not how the language works, and that would break substitutability. A `Base` is not usable wherever a `Derived` is expected, because it might not be a `Derived`.

**Comment by CaseyCarter** [comment]
> `Derived` cannot be the common type between `Derived` and `Base` (in general)

Agreed - and that doesn't contradict anything I said above. The two arguments taken together indicate that  `Derived` and `Base` should have either no common type or a common type derived from (or the same as) `Derived`.

In this case `common_XXX` is in fact more-specific than `std::XXX`: it has additional `common_reference` behavior. The implicit conversion from `common_XXX` to `std::XXX` does lose information, which we allow with full knowledge for backwards compatibility.

**Comment by ericniebler** [comment]
> `Derived` and `Base` should have either no common type or a common type derived from (or the same as) `Derived`.

Given:

```
struct Base {};
struct Derived : Base {};
```

What is the type of `true ? Base{} : Derived{}`? Answer: `Base`. What is the type `std::common_type_t<Base, Derived>`? Answer: `Base`. Are you suggesting we make `common_type` inconsistent with the ternary conditional operator *and* break back-compat for `common_type`?

**Comment by CaseyCarter** [comment]
That example is the trivial case I described above: it is semantically sound for `Base` to be the common type of `Base` and `Derived` when the two types are precisely equivalent.  You're right, though - I'm overreaching. If the conversion from `T` and `U` (resp. `U` to `T`) is one-to-one, than `Common<T, U>` is satisfied and `U` (resp. `T`) is a common type of `T` and `U`. 

The weird thing here is that the conversion from `Derived` to `Base` *is* one-to-one. (The warning presumably exists because conversion-to-base is predominantly a lossy conversion.) Typically we would add behaviors to `Base` rather than "renaming" to `Derived` and adding behavior to that. It's necessary in this case because we can't always add the behaviors we want to types in namespace `std`.

In any case, I see no reason to *prefer* `std::XXX` as the common type of `std::XXX` and `ranges::common_XXX` when `ranges::common_XXX` is an equally viable common type and won't trigger this warning.

**Comment by ericniebler** [comment]
> In any case, I see no reason to *prefer* `std::XXX` as the common type of `std::XXX` and `ranges::common_XXX` when `ranges::common_XXX` is an equally viable common type and won't trigger this warning.

There *is* a reason, or there was, I just can't remember it. You could try making the change and seeing what breaks. Then you can do both of us a favor and comment the code. (I could do all this too, I'm just less motivated since this doesn't offend my sensibilities.)

**Comment by CaseyCarter** [comment]
> There *is* a reason, or there was, I just can't remember it. You could try making the change and seeing what breaks. Then you can do both of us a favor and comment the code.

That's the smartest thing yet posted in this thread. I'm sure we've been over this before - at least in passing - but I can't remember what conclusion we reached if any. It would be nice to tie up the loose end with a comment in the code that will stop me the next time I rediscover this "problem."
